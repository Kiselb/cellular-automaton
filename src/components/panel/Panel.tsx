import React from "react";
import "./panel.css";
import { Cell, CellParams } from "../cell/Cell";

interface PanelProps {
  data: number[][];
  onChange: (params: CellParams) => void;
}

export const Panel = ({ data, onChange, ...props }: PanelProps) => {
  const cellOnClick = (params: CellParams) => {
    onChange(params);
  };
  return (
    <div className={["panel"].join(" ")} {...props}>
      {data.map((row, rowindex) =>
        row.map((item, colindex) => (
          <div
            key={`Wrap:R${rowindex}:C${colindex}`}
            className={[
              "grid",
              "grid--cell",
              colindex % row.length === 0 ? "grid--wrap" : "",
            ].join(" ")}
            data-testid="Panel"
          >
            <Cell
              data-testid={`R${rowindex}:C${colindex}`}
              row={rowindex}
              col={colindex}
              generation={item}
              onClick={cellOnClick}
              colorEmpty="#ececec"
              colorGamma="#77FF77"
            ></Cell>
          </div>
        ))
      )}
    </div>
  );
};
