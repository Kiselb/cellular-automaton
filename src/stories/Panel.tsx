import React from "react";

import "./panel.css";
import { Cell, CellStatus, CellParams } from "./Cell";

export type PanelMode = "display" | "paused";

interface PanelProps {
  data: CellStatus[][];
  mode: PanelMode;
  onChange: (params: CellParams) => void;
}

export const Panel = ({
  data,
  mode: PanelMode = "paused",
  onChange,
  ...props
}: PanelProps) => {
  const cellOnClick = (params: CellParams) => {
    onChange(params);
  };
  return (
    <div className={["panel"].join(" ")} {...props}>
      {data.map((row, rowindex) =>
        row.map((item, colindex) => (
          <div
            className={[
              "grid",
              "grid--cell",
              colindex % row.length === 0 ? "grid--wrap" : "",
            ].join(" ")}
          >
            <Cell
              row={rowindex}
              col={colindex}
              status={item}
              onClick={cellOnClick}
            ></Cell>
          </div>
        ))
      )}
    </div>
  );
};
