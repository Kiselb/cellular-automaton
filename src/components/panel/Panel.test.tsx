import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";

import "@testing-library/jest-dom";
import { Panel } from "./Panel";

it("Panel renders correctly", () => {
  const data: number[][] = Array.from({ length: 4 }, () =>
    Array.from({ length: 4 }, () => 0)
  );
  const { asFragment } = render(
    <Panel data={data} onChange={() => console.log("OK")} />
  );
  screen.debug();
  expect(asFragment()).toMatchSnapshot();
});
it("Panel Cell click event", () => {
  const data: number[][] = Array.from({ length: 4 }, () =>
    Array.from({ length: 4 }, () => 0)
  );
  const probe = jest.fn();
  render(<Panel data={data} onChange={probe} />);
  const div = screen.getByTestId("R1:C1");
  fireEvent.click(div);
  expect(probe).toBeCalled();
});
