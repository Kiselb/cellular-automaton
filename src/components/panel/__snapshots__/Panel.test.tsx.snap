// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="grid--row"
    >
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R0:C0"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R0:C1"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R0:C2"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R0:C3"
          style="background: rgb(236, 236, 236);"
        />
      </div>
    </div>
    <div
      class="grid--row"
    >
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R1:C0"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R1:C1"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R1:C2"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R1:C3"
          style="background: rgb(236, 236, 236);"
        />
      </div>
    </div>
    <div
      class="grid--row"
    >
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R2:C0"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R2:C1"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R2:C2"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R2:C3"
          style="background: rgb(236, 236, 236);"
        />
      </div>
    </div>
    <div
      class="grid--row"
    >
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R3:C0"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R3:C1"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R3:C2"
          style="background: rgb(236, 236, 236);"
        />
      </div>
      <div
        class="grid grid--cell"
        data-testid="Panel"
      >
        <div
          class="cell"
          data-testid="R3:C3"
          style="background: rgb(236, 236, 236);"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
